vignettes/tcplfit2-vignette.Rmd

```diff
@@ -161,3 +161,147 @@ The input data for this example contains 6 signatures for one chemical in a tran
+setkeyv(dat, c('acid', 'srcf', 'apid', 'coli', 'rowi', 'spid', 'conc'))
+# Define rpid column for test compound wells
+nconc <- dat[wllt == "t" , ## denotes test well as the well type (wllt)
+             list(n = lu(conc)),
+             by = list(acid, apid, spid)][ , list(nconc = min(n)), by = acid]
+dat[wllt == "t" & acid %in% nconc[nconc > 1, acid],
+    rpid := paste(acid, spid, wllt, srcf, apid, "rep1", conc, sep = "_")]
+dat[wllt == "t" & acid %in% nconc[nconc == 1, acid],
+    rpid := paste(acid, spid, wllt, srcf, "rep1", conc, sep = "_")]
+# Define rpid column for non-test compound wells
+dat[wllt != "t",
+    rpid := paste(acid, spid, wllt, srcf, apid, "rep1", conc, sep = "_")]
+
+# set repid based on rowid
+dat[, dat_rpid := rowid(rpid)]
+dat[, rpid := sub("_rep[0-9]+.*", "",rpid, useBytes = TRUE)]
+dat[, rpid := paste0(rpid,"_rep",dat_rpid)]
+
+## Define concentration index
+indexfunc <- function(x) as.integer(rank(unique(x))[match(x, unique(x))])
+dat[ , cndx := indexfunc(conc), by = list(rpid)]
+```
+
+Once the concentration index has been calculated, the level 4 methods can be loaded in a similar way to level 3.
+
+```{r example4_mthdlist_l4, fig.height = 6, fig.width = 7}
+mthd_funcs_l4 <- tcpl:::mc4_mthds()
+DT::datatable(tcpl::tcplMthdList(4),options = list(scrollX = T))
+```
+
+There are much fewer level 4 methods, but generally it is a requirement to assign a method that calculates the bmad and assign a method that calculates the standard deviation of the noise.
+
+```{r example4_mc4methods, fig.height = 6, fig.width = 7, results = 'hide'}
+# apply level 4 methods
+lapply(mthd_funcs_l4[["bmad.aeid.lowconc.twells"]](),eval)
+lapply(mthd_funcs_l4[["onesd.aeid.lowconc.twells"]](),eval)
+lapply(mthd_funcs_l4[["bidirectional.false"]](),eval)
+
+```
+
+After all methods up to level 4 have been applied, the model fitting steps can begin.  In tcpl this would be considered level 4, and is where `tcplfit2` is used to fit all of the models as a dependency of tcpl.
+
+
+```{r example4_fitting, fig.height = 6, fig.width = 7}
+
+#do tcplfit2 fitting
+
+myfun <- function(y) {
+  res <- tcplfit2::tcplfit2_core(y$conc,
+                          y$resp,
+                          cutoff = unique(y$bmad),
+                          bidirectional = TRUE,
+                          verbose = FALSE,
+                          force.fit = TRUE,
+                          fitmodels = c("cnst", "hill", "gnls", "poly1",
+                                        "poly2", "pow", "exp2", "exp3",
+                                        "exp4", "exp5")
+                          )
+  list(list(res)) #use list twice because data.table uses list(.) to look for values to assign to columns
+}
+
+
+# only want to run tcplfit2 for test wells in this case
+dat[wllt == 't',params:= myfun(.SD), by = .(spid)]
```

Comment:
At line 281 I wasn't sure what was happening exactly - would it be possibly to explain a bit more in comment?